external/kafka-0-10-sql/src/main/scala/ContinuousKafkaReader.scala

```diff
@@ -0,0 +1,246 @@
+        case SpecificOffsetRangeLimit(p) => fetchAndVerify(p)
+      }
+      logInfo(s"Initial offsets: $offsets")
+      offsets
+    }
+  }
+
+  override def getStartOffset(): Offset = offset
+
+  override def deserializeOffset(json: String): Offset = {
+    KafkaSourceOffset(JsonUtils.partitionOffsets(json))
+  }
+
+  override def createUnsafeRowReadTasks(): java.util.List[ReadTask[UnsafeRow]] = {
+    import scala.collection.JavaConverters._
+
+    val oldStartOffsets = KafkaSourceOffset.getPartitionOffsets(offset)
+
+    val newPartitions =
+      kafkaReader.fetchLatestOffsets().keySet.diff(oldStartOffsets.keySet)
+    val newPartitionOffsets = kafkaReader.fetchEarliestOffsets(newPartitions.toSeq)
+    val startOffsets = oldStartOffsets ++ newPartitionOffsets
+
+    knownPartitions = startOffsets.keySet
+
+    startOffsets.toSeq.map {
+      case (topicPartition, start) =>
+        ContinuousKafkaReadTask(
+          topicPartition, start, executorKafkaParams, pollTimeoutMs, failOnDataLoss)
+          .asInstanceOf[ReadTask[UnsafeRow]]
+    }.asJava
+  }
+
+  /** Stop this source and free any resources it has allocated. */
+  def stop(): Unit = synchronized {
+    kafkaReader.close()
+  }
+
+  override def commit(end: Offset): Unit = {}
+
+  override def needsReconfiguration(): Boolean = {
+    knownPartitions != null && kafkaReader.fetchLatestOffsets().keySet != knownPartitions
+  }
+
+  override def toString(): String = s"KafkaSource[$kafkaReader]"
+
+  /**
+   * If `failOnDataLoss` is true, this method will throw an `IllegalStateException`.
+   * Otherwise, just log a warning.
+   */
+  private def reportDataLoss(message: String): Unit = {
+    if (failOnDataLoss) {
+      throw new IllegalStateException(message + s". $INSTRUCTION_FOR_FAIL_ON_DATA_LOSS_TRUE")
+    } else {
+      logWarning(message + s". $INSTRUCTION_FOR_FAIL_ON_DATA_LOSS_FALSE")
+    }
+  }
+}
+
+case class ContinuousKafkaReadTask(
+    topicPartition: TopicPartition,
+    start: Long,
+    kafkaParams: java.util.Map[String, Object],
+    pollTimeoutMs: Long,
+    failOnDataLoss: Boolean)
+  extends ReadTask[UnsafeRow] {
+  override def createDataReader(): ContinuousKafkaDataReader = {
+    new ContinuousKafkaDataReader(topicPartition, start, kafkaParams, pollTimeoutMs, failOnDataLoss)
+  }
+}
+
+class ContinuousKafkaDataReader(
+    topicPartition: TopicPartition,
+    start: Long,
+    kafkaParams: java.util.Map[String, Object],
+    pollTimeoutMs: Long,
+    failOnDataLoss: Boolean)
+  extends ContinuousDataReader[UnsafeRow] {
+  private val topic = topicPartition.topic
+  private val kafkaPartition = topicPartition.partition
+  private val consumer = CachedKafkaConsumer.createUncached(topic, kafkaPartition, kafkaParams)
+
+  private val closed = new AtomicBoolean(false)
+
+  private var nextKafkaOffset = start match {
+    case s if s >= 0 => s
+    case KafkaOffsetRangeLimit.EARLIEST => consumer.getAvailableOffsetRange().earliest
```

Comment:
I think this resolution actually doesn't do anything at all, because the offsets were already resolved earlier. I think I had the resolution only here originally, and forgot to delete it because it ends up as a no-op. I'll remove it now.